Add LDSR and "GoLatent?" Upscaling

d8ahazard opened the pull request:
Add a basic implementation of Latent Diffusion Super Resizing, based off of the mod by @Hafiidz.

https://github.com/Hafiidz/latent-diffusion

Right now, I've added it as an upscaler in the extras section, but I'm not sure if that's the ideal place for it? 

Either way, currently, it doesn't have hooks for all of the available params in the SuperResizer, which are total steps, scale to downscale before adding details, and the scale to go back to after doing the detailing?

So, with a little help from @AUTOMATIC1111 on the UI side, I think this would be a *great* addition to sd-webui. I think I found where in the original code there's an option to set how big to resize, but I'd have to do more experimentation. I have a fork of Hafiidz's project that I'm going to do some code cleanup on, maybe see if I can better expose the scale settings.

I should point out that it takes a lot longer to resize than using other methods - around 4 minutes, but I think the results are WELL worth it. It adds a lot of detail and cleans stuff up without changing the composition of the photo at all. :D 

![LDSR_Compare](https://user-images.githubusercontent.com/1633844/191326375-0ad38a66-9951-4dca-8242-da02ed0430cc.png)

d8ahazard committed:
Add LDSR Upscaling

devilismyfriend commented:
Nice to see my implementation being moved here, I love it as well!

d8ahazard commented:
> Nice to see my implementation being moved here, I love it as well!, keep in mind the upscaling factor inside LDSR.py is broken as of right now.

I just submitted a PR to the fork of your fork to do some major cleanup - if you want to point me at the lines containing the upscaling factor, I could maybe take a look and try fixing it. ;)

AUTOMATIC1111 commented:
There are some prints that probably should be removed (because they would print the same thing every time and not be informative). As for settings, you can add whatever options you like to `opts` in `shared` and read them when you are upscaling the image.  It would be preferable to download and load that model when you are trying to upscale an image rather than when you are launching the program. Additionally, please test that it works with full precision, just in case.

d8ahazard committed:
Remove pointless debugging, setup on upscale

Remove debugging messages.
Download pre-trained models on first upscale, versus program launch.

AUTOMATIC1111 commented:
Well, you can just put options for the upscaler into the setting screens, not into the two screens you listed. This would be a lot more simple in implementation and i really like simple.

Full precision is enabled by running with `--precision full --no-half`.

d8ahazard commented:
> Well, you can just put options for the upscaler into the setting screens, not into the two screens you listed. This would be a lot more simple in implementation and i really like simple.

I'm also a fan of simple. I'll add those and test precision shortly.

d8ahazard committed:
Ensure ldsr repo is up to date on launch?

d8ahazard committed:
Enable settings integration

d8ahazard committed:
Add settings integration to shared settings

d8ahazard committed:
Merge branch 'master' into LDSR

d8ahazard committed:
We probably want to still load the LSDR module...

d8ahazard committed:
Merge branch 'LDSR' of https://github.com/d8ahazard/stable-diffusion-webui into LDSR

d8ahazard commented:
I have added settings in the proper place, switched the branch we're pulling from to my fork for the time being, as I've made changes to the base LDSR class that allows for int values versus strings for the settings, and re-added a proper initializer to add the module to the list of shared scalers.

Just waiting for the checkpoint to re-download, and I can confirm that it works in full precision.

For my own knowledge, what's the difference with full precision, and is it something I'd want to use regularly?

C43H66N12O12S2 commented:
Full precision is FP32, half precision is FP16. In theory, FP16 is less precise and can result in lower quality but I can't remember the last time I saw this happen.

Most people who use FP32 are those who are forced to.

d8ahazard committed:
That's how that works...

d8ahazard committed:
I can't read, or my IDE sucks.

d8ahazard commented:
Mr Whiskers, in glorious FP32. :P 

![00006](https://user-images.githubusercontent.com/1633844/191376210-06a42199-2e99-4363-ad01-1017b2498258.png)

d8ahazard commented:
Processed at 50 steps instead of 100, and with 1/3 scale options for both up and down. Was crazy fast, but not sure how the details compare.

![00007](https://user-images.githubusercontent.com/1633844/191376417-b167c088-8256-44b6-99d7-44fef76c9a64.png)

Hafiidz commented:
> > Nice to see my implementation being moved here, I love it as well!, keep in mind the upscaling factor inside LDSR.py is broken as of right now.
> 
> I just submitted a PR to the fork of your fork to do some major cleanup - if you want to point me at the lines containing the upscaling factor, I could maybe take a look and try fixing it. ;)

PR approved, thanks. I have tested with alternate UI, working well as well.

ArrowM commented:
I think modules/ldsr_model.py:75
`image = LDSR_obj.super_resolution(image, ddim_steps, pre_scale, post_scale)`
is meant to be 
`image = LDSR_obj.superResolution(image, ddim_steps, pre_scale, post_scale)`

d8ahazard commented:
> I think modules/ldsr_model.py:75 `image = LDSR_obj.super_resolution(image, ddim_steps, pre_scale, post_scale)` is meant to be `image = LDSR_obj.superResolution(image, ddim_steps, pre_scale, post_scale)`

No, I added a renamed/overflow method that takes integers instead of strings to the LDSR class, this way, the original superResolution method still works, but we can use the new one with the settings from our UI.

ArrowM commented:
> No, I added a renamed/overflow method that takes integers instead of strings to the LDSR class, this way, the original superResolution method still works, but we can use the new one with the settings from our UI.
I'm on the latest version of master (which is set to your latent-diffusion repo). Maybe I'm missing something, but I get the following error when I run it without renaming to "superResolution". 
`Error completing request
Arguments: (<PIL.Image.Image image mode=RGB size=512x512 at 0x1FA80F1D180>, None, 0, 0, 0, 4, 12, 0, 1) {}
Traceback (most recent call last):
  File "C:\Users\Dalton\source\repos\stable-diffusion-webui-auto\modules\ui.py", line 135, in f
    res = list(func(*args, **kwargs))
  File "C:\Users\Dalton\source\repos\stable-diffusion-webui-auto\webui.py", line 58, in f
    res = func(*args, **kwargs)
  File "C:\Users\Dalton\source\repos\stable-diffusion-webui-auto\modules\extras.py", line 87, in run_extras
    res = upscale(image, extras_upscaler_1, upscaling_resize)
  File "C:\Users\Dalton\source\repos\stable-diffusion-webui-auto\modules\extras.py", line 81, in upscale
    c = upscaler.upscale(image, image.width * resize, image.height * resize)
  File "C:\Users\Dalton\source\repos\stable-diffusion-webui-auto\modules\images.py", line 404, in upscale
    img = self.do_upscale(img)
  File "C:\Users\Dalton\source\repos\stable-diffusion-webui-auto\modules\ldsr_model.py", line 25, in do_upscale
    return upscale_with_ldsr(img)
  File "C:\Users\Dalton\source\repos\stable-diffusion-webui-auto\modules\ldsr_model.py", line 75, in upscale_with_ldsr
    image = LDSR_obj.super_resolution(image, ddim_steps, pre_scale, post_scale)
AttributeError: 'LDSR' object has no attribute 'super_resolution'`

AUTOMATIC1111 commented:
All right, get those `.idea` files out of the pr.

d8ahazard committed:
Too tired to commit stuff

d8ahazard committed:
Update .gitignore

d8ahazard committed:
Revert ld URL.

d8ahazard committed:
Merge branch 'master' into LDSR

d8ahazard commented:
> All right, get those `.idea` files out of the pr.

I'll do you one better and just ignore the whole folder, as I'm sure it's likely to happen in the future otherwise.

AUTOMATIC1111 commented:
there is `.git/info/exclude`, you know

AUTOMATIC1111 commented:
So the status so far is,
- it fails to download model/config (exception below)
- it does tell you where to put those so if you download them yourself, that can be worked around
- the function LDSR_obj.super_resolution as arrow, pointed, does not exist
- if you replace it with LDSR_obj.superResolution, it finally works.

Here's original pic upscaled with box:
![00156-2756194963](https://user-images.githubusercontent.com/20920490/191517447-80fd3539-af5f-4f16-9c7d-f70a265c3de5.png)
ESRGAN's 4x_foolhardy_Remacri:
![02047](https://user-images.githubusercontent.com/20920490/191517553-11c1d7c2-a968-4e92-9445-4e569663e75b.png)
LDSR:
![02048](https://user-images.githubusercontent.com/20920490/191517578-00b69e13-c30e-4728-8a48-df9076a59356.png)

```
Error importing LDSR:
Traceback (most recent call last):
  File "B:\soft\Python310\lib\urllib\request.py", line 1348, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "B:\soft\Python310\lib\http\client.py", line 1282, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "B:\soft\Python310\lib\http\client.py", line 1328, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "B:\soft\Python310\lib\http\client.py", line 1277, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "B:\soft\Python310\lib\http\client.py", line 1037, in _send_output
    self.send(msg)
  File "B:\soft\Python310\lib\http\client.py", line 975, in send
    self.connect()
  File "B:\soft\Python310\lib\http\client.py", line 1454, in connect
    self.sock = self._context.wrap_socket(self.sock,
  File "B:\soft\Python310\lib\ssl.py", line 513, in wrap_socket
    return self.sslsocket_class._create(
  File "B:\soft\Python310\lib\ssl.py", line 1071, in _create
    self.do_handshake()
  File "B:\soft\Python310\lib\ssl.py", line 1342, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:997)
```

devilismyfriend commented:
> > Nice to see my implementation being moved here, I love it as well!, keep in mind the upscaling factor inside LDSR.py is broken as of right now.
> 
> I just submitted a PR to the fork of your fork to do some major cleanup - if you want to point me at the lines containing the upscaling factor, I could maybe take a look and try fixing it. ;)

I've been away from my PC for around 3 weeks now so it's been difficult to follow along, but it's v_f or something like that.

Also might want to expose the steps control to the users in the Extras tab, 100 is the default but I found 50 to be close enough most of the time, could also expose the downsampling option.
